src/gui/app.cpp

```diff
@@ -11,8 +11,148 @@
+	};
+
+	ConnectorInitArgs(const QCommandLineParser& parser, QStringList nvimArgs) noexcept
+		: type{ getConnectorType(parser) }
+		, timeout{ parser.value("timeout").toInt() }
+		, server{ parser.value("server") }
+		, nvim{ parser.value("nvim") }
+		, positionalArgs{ parser.positionalArguments() }
+		, neovimArgs{ std::move(nvimArgs) }
+	{
+
+	}
+
+	ConnectorInitArgs(
+		Type _type, int _timeout, QString _server, QString _nvim, QStringList nvimArgs) noexcept
+		: type{ _type }
+		, timeout{ _timeout }
+		, server{ std::move(_server) }
+		, nvim{ std::move(_nvim) }
+		, positionalArgs{}
+		, neovimArgs{std::move(nvimArgs)}
+	{
+	}
+
+	const Type type;
+	const int timeout;
+	const QString server;
+	const QString nvim;
+	const QStringList positionalArgs;
+	const QStringList neovimArgs;
+
+	Type getConnectorType(const QCommandLineParser& parser) noexcept
+	{
+		if (parser.isSet("server")) {
+			return ConnectorInitArgs::Type::Server;
+		}
+
+		if (parser.isSet("embed")) {
+			return ConnectorInitArgs::Type::Embed;
+		}
+
+		if (parser.isSet("spawn") && !parser.positionalArguments().isEmpty()) {
+			return ConnectorInitArgs::Type::Spawn;
+		}
+
+		return ConnectorInitArgs::Type::Default;
+	}
+};
+
+NeovimConnector* connectToRemoteNeovim(const ConnectorInitArgs& args) noexcept
+{
+	NeovimConnector *connector{nullptr};
+	if (args.type == ConnectorInitArgs::Type::Embed) {
+		connector = NeovimQt::NeovimConnector::fromStdinOut();
+	}
+
+	if (args.type == ConnectorInitArgs::Type::Server) {
+		connector = NeovimQt::NeovimConnector::connectToNeovim(args.server);
+	}
+
+	if (args.type == ConnectorInitArgs::Type::Spawn) {
+		Q_ASSERT(!args.positionalArgs.isEmpty());
+		connector = NeovimQt::NeovimConnector::spawn(args.positionalArgs.mid(1),
+				args.positionalArgs.at(0));
+	}
+
+	if (!connector) {
+		connector = NeovimQt::NeovimConnector::spawn(
+				args.neovimArgs + args.positionalArgs, args.nvim);
+	}
+
+	connector->setRequestTimeout(args.timeout);
+	return connector;
+}
+
+void onWindowActiveChanged(MainWindow& window) noexcept
+{
+	s_lastActiveWindow = &window;
+}
+
+void onFileOpenEvent(const QList<QUrl>& files) noexcept
+{
+	Q_ASSERT(s_lastActiveWindow);
+	s_lastActiveWindow->shell()->openFiles(files);
+}
+
+void onWindowClosing(int status) noexcept
+{
+	s_exitStatus = status;
+}
+
+void onWindowDestroyed() noexcept
+{
+	if (s_windows.empty()) {
+		qApp->exit(s_exitStatus);
+	}
+}
+
+MainWindow* createWindow(NeovimConnector* connector) noexcept
+{
+	Q_ASSERT(connector);
+	NeovimQt::MainWindow *win = new NeovimQt::MainWindow(connector);
+	win->setAttribute(Qt::WA_DeleteOnClose);
+
+	App *app = qobject_cast<App *>(App::instance());
+	Q_ASSERT(app);
+
+	QObject::connect(win, &MainWindow::closing, app, onWindowClosing);
+    QObject::connect(win, &MainWindow::destroyed, app, [win]() {
+        auto foundIt = std::find_if(
+            s_windows.cbegin(), s_windows.cend(),
+            [win](const std::reference_wrapper<NeovimQt::MainWindow>& rw) {
```

Comment:
Curious... Why `reference_wrapper` here? Shouldn't a plain `const MainWindow&` work?

Also no need for `NeovimQt`.

Reply:
We can't have vector of references since references are not assignable. That's why the use of `std::reference_wrapper`.